File: packages/sdk-py/gatewerk/resources/reviews.py
"""Reviews resource for the Gatewerk Python SDK."""

from __future__ import annotations

import warnings
from typing import Any, Iterator, Optional

from .._base import BaseResource
from ..types import Review, ReviewList, ReviewVersion

_LEGACY_SUNSET = "2026-12-01"


class ReviewsResource(BaseResource):

    def create(
        self,
        template: str,
        payload: dict[str, Any],
        *,
        callback_url: Optional[str] = None,
        priority: Optional[str] = None,
        actions: Optional[list[str]] = None,
        confidence: Optional[float] = None,
        irreversibility: Optional[str] = None,
        assignee: Optional[str] = None,
        metadata: Optional[dict[str, Any]] = None,
        timeout: Optional[float] = None,
        idempotency_key: Optional[str] = None,
    ) -> Review:
        body: dict[str, Any] = {"template": template, "payload": payload}
        if callback_url is not None:
            body["callback_url"] = callback_url
        if priority is not None:
            body["priority"] = priority
        if actions is not None:
            body["actions"] = actions
        if confidence is not None:
            body["confidence"] = confidence
        if irreversibility is not None:
            body["irreversibility"] = irreversibility
        if assignee is not None:
            body["assignee"] = assignee
        if metadata is not None:
            body["metadata"] = metadata
        if idempotency_key is not None:
            body["idempotency_key"] = idempotency_key

        data = self._request("POST", "/api/v1/reviews", json=body, timeout=timeout)
        return Review.model_validate(data)

    def get(self, review_id: str, *, timeout: Optional[float] = None) -> Review:
        data = self._request("GET", f"/api/v1/reviews/{review_id}", timeout=timeout)
        return Review.model_validate(data)

    def list(
        self,
        *,
        status: Optional[str] = None,
        priority: Optional[str] = None,
        template: Optional[str] = None,
        assignee: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        timeout: Optional[float] = None,
    ) -> ReviewList:
        params: dict[str, Any] = {}
        if status is not None:
            params["status"] = status
        if priority is not None:
            params["priority"] = priority
        if template is not None:
            params["template"] = template
        if assignee is not None:
            params["assignee"] = assignee
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset

        data = self._request("GET", "/api/v1/reviews", params=params, timeout=timeout)
        return ReviewList.model_validate(data)

    def decide(
        self,
        review_id: str,
        decision: str,
        *,
        feedback: Optional[str] = None,
        edited_payload: Optional[dict[str, Any]] = None,
        reviewer: Optional[str] = None,
        prompt_edit: Optional[str] = None,
        version: Optional[int] = None,
        action_value: Optional[str] = None,
        action_label: Optional[str] = None,
        timeout: Optional[float] = None,
    ) -> Review:
        warnings.warn(
            "reviews.decide() is deprecated for session-authenticated callers; "
            "use reviews.action(review_id, action_id=...) instead. "
            "API-key (agent) callers should continue using reviews.decide() until "
            "the action endpoint supports api-key auth. "
            f"Removed in v2.0 (Sunset: {_LEGACY_SUNSET}). "
            "See https://docs.gatewerk.com/migration/configurable-actions",
            DeprecationWarning,
            stacklevel=2,
        )
        body: dict[str, Any] = {"decision": decision}
        if feedback is not None:
            body["feedback"] = feedback
        if edited_payload is not None:
            body["edited_payload"] = edited_payload
        if reviewer is not None:
            body["reviewer"] = reviewer
        if prompt_edit is not None:
            body["prompt_edit"] = prompt_edit
        if version is not None:
            body["version"] = version
        if action_value is not None:
            body["action_value"] = action_value
        if action_label is not None:
            body["action_label"] = action_label

        data = self._request("POST", f"/api/v1/reviews/{review_id}/decide", json=body, timeout=timeout)
        return Review.model_validate(data)

    def action(
        self,
        review_id: str,
        action_id: str,
        *,
        feedback: Optional[str] = None,
        edited_payload: Optional[dict[str, Any]] = None,
        version: Optional[int] = None,
        timeout: Optional[float] = None,
    ) -> Review:
        """Invoke a configurable action on a review.

        Requires session authentication today. Use decide() / retry() / cancel_request()
        for api-key (agent) callers until the action endpoint supports api-key auth.
        """
        body: dict[str, Any] = {"action_id": action_id}
        if feedback is not None:
            body["feedback"] = feedback
        if edited_payload is not None:
            body["edited_payload"] = edited_payload
        if version is not None:
            body["version"] = version

        data = self._request("POST", f"/api/v1/reviews/{review_id}/action", json=body, timeout=timeout)
        return Review.model_validate(data)

    def retry(
        self,
        review_id: str,
        *,
        feedback: Optional[str] = None,
        prompt_edit: Optional[str] = None,
        timeout: Optional[float] = None,
    ) -> Review:
        warnings.warn(
            "reviews.retry() is deprecated for session-authenticated callers; "
            "use reviews.action(review_id, action_id=...) instead. "
            "API-key (agent) callers should continue using reviews.retry() until "
            "the action endpoint supports api-key auth. "
            f"Removed in v2.0 (Sunset: {_LEGACY_SUNSET}). "
            "See https://docs.gatewerk.com/migration/configurable-actions",
            DeprecationWarning,
            stacklevel=2,
        )
        body: dict[str, Any] = {}
        if feedback is not None:
            body["feedback"] = feedback
        if prompt_edit is not None:
            body["prompt_edit"] = prompt_edit

        data = self._request("POST", f"/api/v1/reviews/{review_id}/retry", json=body, timeout=timeout)
        return Review.model_validate(data)

    def update(
        self,
        review_id: str,
        payload: dict[str, Any],
        version: int,
        *,
        timeout: Optional[float] = None,
    ) -> Review:
        body = {"payload": payload, "version": version}
        data = self._request("PUT", f"/api/v1/reviews/{review_id}", json=body, timeout=timeout)
        return Review.model_validate(data)

    def cancel_request(self, review_id: str, *, timeout: Optional[float] = None) -> Review:
        """Revert a 'changes_requested' review back to pending."""
        warnings.warn(
            "reviews.cancel_request() is deprecated for session-authenticated callers; "
            "use reviews.action(review_id, action_id=...) instead. "
            "API-key (agent) callers should continue using reviews.cancel_request() until "
            "the action endpoint supports api-key auth. "
            f"Removed in v2.0 (Sunset: {_LEGACY_SUNSET}). "
            "See https://docs.gatewerk.com/migration/configurable-actions",
            DeprecationWarning,
            stacklevel=2,
        )
        data = self._request("POST", f"/api/v1/reviews/{review_id}/cancel-request", timeout=timeout)
        return Review.model_validate(data)

    def versions(self, review_id: str, *, timeout: Optional[float] = None) -> list[ReviewVersion]:
        """Get version history for a review."""
        data = self._request("GET", f"/api/v1/reviews/{review_id}/versions", timeout=timeout)
        if isinstance(data, list):
            return [ReviewVersion.model_validate(v) for v in data]
        items = data.get("items", data.get("versions", []))
        return [ReviewVersion.model_validate(v) for v in items]

    def create_token(self, review_id: str, *, timeout: Optional[float] = None) -> dict[str, Any]:
        """Generate a shareable review link token."""
        return self._request("POST", f"/api/v1/reviews/{review_id}/token", timeout=timeout)

    def list_auto_paginate(
        self,
        *,
        status: Optional[str] = None,
        priority: Optional[str] = None,
        template: Optional[str] = None,
        assignee: Optional[str] = None,
        batch_size: int = 50,
    ) -> Iterator[Review]:
        """Iterate through all reviews, automatically handling pagination."""
        offset = 0
        while True:
            page = self.list(
                status=status, priority=priority, template=template,
                assignee=assignee, limit=batch_size, offset=offset,
            )
            yield from page.items
            if not page.has_more:
                break
            offset += batch_size

File: packages/sdk-ts/src/resources/reviews.ts
import type { Result } from "../errors.js";
import { retryingRequest } from "../http.js";
import type { ReviewActionBody } from "@gatewerk/shared";

// Industry-standard SDK deprecation pattern: emit runtime warning exactly
// once per method per process, in addition to @deprecated JSDoc which IDEs
// surface at edit time. Mirrors Stripe/OpenAI/Twilio Node SDKs.
const __deprecationWarned = new Set<string>();
function warnDeprecated(method: string, successor: string, sunset: string): void {
  if (__deprecationWarned.has(method)) return;
  __deprecationWarned.add(method);
  // eslint-disable-next-line no-console
  console.warn(
    `[gatewerk] reviews.${method}() is deprecated for session-authenticated callers; ` +
    `use reviews.${successor}() instead. API-key (agent) callers should continue ` +
    `using reviews.${method}() until the action endpoint supports api-key auth. ` +
    `Removed in v2.0 (Sunset: ${sunset}). ` +
    `See https://docs.gatewerk.com/migration/configurable-actions for migration.`,
  );
}
const LEGACY_SUNSET = "2026-12-01";

export interface CreateReviewInput {
  template: string;
  payload: Record<string, unknown>;
  callback_url?: string;
  priority?: string;
  actions?: string[];
  confidence?: number;
  irreversibility?: string;
  assignee?: string;
  metadata?: Record<string, unknown>;
  timeout?: { action: string; seconds: number };
  idempotency_key?: string;
}

export interface DecideInput {
  decision: string;
  feedback?: string;
  edited_payload?: Record<string, unknown>;
  reviewer?: string;
  prompt_edit?: string;
  // Wave 2: surface every field the backend's ReviewDecideBodySchema accepts
  // (apps/api/src/routes/reviews/decide.ts → ReviewDecideBodySchema). `version`
  // is the optimistic-concurrency guard for edited_payload submissions;
  // `action_value` + `action_label` carry the chosen action when the template
  // exposes a non-default action set (M11).
  version?: number;
  action_value?: string;
  action_label?: string;
}

export type ActionInput = ReviewActionBody;

export interface ListFilters {
  status?: string;
  priority?: string;
  template?: string;
  assignee?: string;
  limit?: number;
  offset?: number;
}

export interface UpdateVersionInput {
  payload: Record<string, unknown>;
  version: number;
}

export interface ReviewTemplateField {
  name: string;
  label: string;
  type: string;
  editable: boolean;
  options?: string[];
}

export interface ReviewTemplate {
  name: string;
  fields: ReviewTemplateField[];
  actions: string[];
}

export interface ReviewDetail {
  id: string;
  object: string;
  project_id: string;
  template_id: string;
  template_slug: string;
  payload: Record<string, unknown>;
  status: string;
  priority: string;
  actions: string[];
  decision?: string;
  feedback?: string;
  edited_payload?: Record<string, unknown>;
  decided_by?: string;
  decided_at?: string;
  /** Monotonically increasing revision counter. Starts at 1. */
  current_version?: number;
  /**
   * Number of revision rounds this review has gone through.
   * Derived: current_version - 1. Always present on read/mutation responses.
   * Agents use this to track how many retries occurred before a decision.
   */
  iteration_count?: number;
  created_at: string;
  updated_at: string;
  template?: ReviewTemplate;
}

export interface ReviewListResult {
  items: ReviewDetail[];
  total: number;
  has_more: boolean;
}

export interface ReviewVersion {
  id: string;
  review_id: string;
  version: number;
  payload: Record<string, unknown>;
  edited_by: string | null;
  created_at: string;
}

export interface ReviewVersionsResult {
  items: ReviewVersion[];
}

export interface ReviewToken {
  token: string;
  url: string;
  expires_at: string;
}


export class ReviewsResource {
  constructor(
    private baseUrl: string,
    private headers: () => Record<string, string>,
  ) {}

  // Retry helper centralised in ../http.ts. Handles
  // {429, 500, 502, 503, 504} + network errors with exponential backoff,
  // honors Retry-After on 429. Mirrors sdk-py BaseResource._request.
  private async request<T>(url: string, init?: RequestInit): Promise<Result<T>> {
    return retryingRequest<T>(url, init, this.headers);
  }

  async create(input: CreateReviewInput): Promise<Result<Record<string, unknown>>> {
    return this.request(`${this.baseUrl}/api/v1/reviews`, {
      method: "POST",
      body: JSON.stringify(input),
    });
  }

  async get(id: string): Promise<Result<ReviewDetail>> {
    return this.request(`${this.baseUrl}/api/v1/reviews/${id}`, { method: "GET" });
  }

  async list(filters?: ListFilters): Promise<Result<ReviewListResult>> {
    const params = new URLSearchParams();
    if (filters?.status) params.set("status", filters.status);
    if (filters?.priority) params.set("priority", filters.priority);
    if (filters?.template) params.set("template", filters.template);
    if (filters?.assignee) params.set("assignee", filters.assignee);
    if (filters?.limit !== undefined) params.set("limit", String(filters.limit));
    if (filters?.offset !== undefined) params.set("offset", String(filters.offset));

    const qs = params.toString();
    return this.request(`${this.baseUrl}/api/v1/reviews${qs ? `?${qs}` : ""}`, { method: "GET" });
  }

  /** @deprecated Use reviews.action(id, { action_id: "approve" | "reject", ... }) instead. Removed in v2.0 (Sunset: 2026-12-01). */
  async decide(id: string, input: DecideInput): Promise<Result<Record<string, unknown>>> {
    warnDeprecated("decide", "action", LEGACY_SUNSET);
    return this.request(`${this.baseUrl}/api/v1/reviews/${id}/decide`, {
      method: "POST",
      body: JSON.stringify(input),
    });
  }

  /**
   * Invoke a configurable action on a review.
   *
   * Requires session authentication today. API-key (agent) callers should
   * continue using decide() / retry() / cancelRequest() until the action
   * endpoint supports api-key auth. POST /api/v1/reviews/:id/action.
   */
  async action(id: string, input: ActionInput): Promise<Result<Record<string, unknown>>> {
    return this.request(`${this.baseUrl}/api/v1/reviews/${id}/action`, {
      method: "POST",
      body: JSON.stringify(input),
    });
  }

  /** @deprecated Use reviews.action(id, { action_id: "request_changes", feedback, ... }) instead. Removed in v2.0 (Sunset: 2026-12-01). */
  async retry(id: string, input: { feedback?: string; prompt_edit?: string }): Promise<Result<Record<string, unknown>>> {
    warnDeprecated("retry", "action", LEGACY_SUNSET);
    return this.request(`${this.baseUrl}/api/v1/reviews/${id}/retry`, {
      method: "POST",
      body: JSON.stringify(input),
    });
  }

  async update(id: string, input: UpdateVersionInput): Promise<Result<Record<string, unknown>>> {
    return this.request(`${this.baseUrl}/api/v1/reviews/${id}`, {
      method: "PUT",
      body: JSON.stringify(input),
    });
  }

  // Wave 2: parity with sdk-py ReviewsResource.cancel_request.
  // POST /api/v1/reviews/:id/cancel-request — reverts a `changes_requested`
  // review to `pending`. See apps/api/src/routes/reviews/decide.ts.
  /** @deprecated Use reviews.action(id, { action_id: "cancel_iteration" }) instead. Removed in v2.0 (Sunset: 2026-12-01). */
  async cancelRequest(id: string): Promise<Result<Record<string, unknown>>> {
    warnDeprecated("cancelRequest", "action", LEGACY_SUNSET);
    return this.request(`${this.baseUrl}/api/v1/reviews/${id}/cancel-request`, {
      method: "POST",
      body: JSON.stringify({}),
    });
  }

  // Wave 2: parity with sdk-py ReviewsResource.versions.
  // GET /api/v1/reviews/:id/versions — lists `review_versions` rows ordered
  // by version desc. See apps/api/src/routes/reviews/lifecycle.ts.
  async versions(id: string): Promise<Result<ReviewVersionsResult>> {
    return this.request(`${this.baseUrl}/api/v1/reviews/${id}/versions`, { method: "GET" });
  }

  // Wave 2: parity with sdk-py ReviewsResource.create_token.
  // POST /api/v1/reviews/:id/token — generates a shareable review-link token.
  // Optional `expiryHours` body field overrides the default per-template expiry.
  // See apps/api/src/routes/reviews/tokens.ts → ReviewTokenBodySchema.
  async createToken(id: string, options?: { expiryHours?: number }): Promise<Result<ReviewToken>> {
    return this.request(`${this.baseUrl}/api/v1/reviews/${id}/token`, {
      method: "POST",
      body: JSON.stringify(options ?? {}),
    });
  }
}
